### PR TITLE
feature(deisctl) allow for graceful in-place upgrades

### PR DESCRIPTION
**Pull request by iancoffey**
First pass at adding upgrade functionality to deisctl.

The whole thing needs :eyes:, but the following things especially.
- [ ] The service keys are persisted by a workaround which resets them with a 30 minute TTL. Is this the best way to keep the services up while the publishers cycle?
- [x] TestUpgradeTakeover fails due to the etcd stuff. If we keep that stuff in there, how should it be tested, a mocked etcd endpoint?
- [x] It would be nice if the command names were better, but its the best Ive come up with so far.
- [x] The docs indicate a deisctl install command that does not yet work - https://github.com/deis/deis.io/pull/114

### CODE REVIEW

**Comment by Joshua-Anderson** [comment]
> TestUpgradeTakeover fails due to the etcd stuff. If we keep that stuff in there, how should it be tested, a mocked etcd endpoint?

I mocked etcd [here](https://github.com/deis/deis/blob/master/deisctl/config/config_test.go), which may be a good starting point. I used an interface to access etcd and then mocked the interface for my tests.


Great Stuff!

File: deisctl/client/client.go [hunk 1]
```diff
@@ -56,6 +59,50 @@ func NewClient(requestedBackend string) (*Client, error) {
...
+func (c *Client) UpgradePrep(argv []string) error {
+	usage := `Prepare platform for graceful upgrade.
+
+Usage:
+  deisctl upgrade-prep [options]
+`
+	_, err := docopt.Parse(usage, argv, true, "", false)
+	if err != nil {
+		return err
+	}
```
Comment (Joshua-Anderson) [inline deisctl/client/client.go]: I think deis tries to go with the 

``` go
if _, err := foo(); err != nil {
  return err
}
```
syntax when possible, though admitingly I may have introduced some of this before I knew the coding style myself. :smiley:

File: deisctl/cmd/cmd_test.go [hunk 2]
```diff
@@ -260,6 +265,55 @@ func TestStartSwarm(t *testing.T) {
...
+	t.Parallel()
+
+	b := backendStub{}
+	expectedRestarted := []string{"router"}
+	expectedStarted := []string{"database", "registry@*", "controller", "builder", "logger", "logspout", "store-volume",
+		"store-gateway@*", "store-metadata", "store-daemon", "store-monitor", "publisher", "router"}
+
+	err := UpgradeTakeover(&b)
+	if err != nil {
+		fmt.Printf("ERRERRERR %s", err)
```
Comment (carmstrong) [inline deisctl/cmd/cmd_test.go]: Is this intentional, or leftover cruft from debugging?

File: deisctl/cmd/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	wg.Wait()
+
+	b.Stop([]string{"store-monitor"}, &wg, Stdout, Stderr)
+	wg.Wait()
+	b.Destroy([]string{"store-monitor"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	fmt.Fprintln(Stdout, "Backend services stopped, but your services should still be serving traffic as normal.")
+
+	fmt.Fprintln(Stdout, "Your cluster is now ready for uprade, use a new deisctl binary and run deisctl take-over to complete this upgrade.")
```
Comment (carmstrong) [inline deisctl/cmd/upgrade.go]: s/uprade/upgrade/

File: deisctl/cmd/cmd_test.go (hunk 2 above)
Comment (iancoffey) [inline deisctl/cmd/cmd_test.go]: oopsies :grin:

File: deisctl/deisctl.go [hunk 4]
```diff
@@ -41,6 +41,9 @@ Commands, use "deisctl help <command>" to learn more:
   refresh-units     refresh unit files from GitHub
   ssh               open an interacive shell on a machine in the cluster
   help              show the help screen for a command
+  upgrade-prep      prepare a running cluster for upgrade
+  upgrade-takeover  allow an upgrade to gracefully takeover a running cluster
+  rolling-restart   perform a rolling restart for instance units
```
Comment (carmstrong) [inline deisctl/deisctl.go]: What's the meaning of "instance unit" here? The term is unfamiliar to me in this context.

Comment (iancoffey) [inline deisctl/deisctl.go]: The term I cribbed from systemds [docs](http://www.freedesktop.org/software/systemd/man/systemd.unit.html) , its just referring to units created like unit@1.service - so the router, registry and gateway are examples. Its intended to indicate that the restart could only function for units created this way - global or one off units couldn't work.

As for better wording - maybe it should say something more generic? "perform a rolling restart for certain valid Deis components" or something.

Comment (carmstrong) [inline deisctl/deisctl.go]: Yeah, maybe something like "perform a rolling restart of a Deis component (currently only router is supported)"

File: deisctl/cmd/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	wg.Wait()
+	b.Destroy([]string{"store-daemon"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	b.Stop([]string{"store-monitor"}, &wg, Stdout, Stderr)
+	wg.Wait()
+	b.Destroy([]string{"store-monitor"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	fmt.Fprintln(Stdout, "Backend services stopped, but your services should still be serving traffic as normal.")
```
Comment (carmstrong) [inline deisctl/cmd/upgrade.go]: "The platform has been stopped, but applications are still serving traffic as normal.

File: deisctl/cmd/upgrade.go (hunk 3 above)
Comment (carmstrong) [inline deisctl/cmd/upgrade.go]: "Your cluster is now ready for upgrade. Install a new deisctl version and run `deisctl take-over`. "

Comment (carmstrong) [inline deisctl/cmd/upgrade.go]: We should also link to the documentation section pertaining to Graceful Upgrades.

**Comment by iancoffey** [comment]
I moved the etcd and etcd mock stuff to its own packages outside of config, since they are not specific to config and other things now need to use them. Im not sure what the best pattern for the namespacing and locations would be, would love input on that and everything else.

**Comment by iancoffey** [comment]
Also, the mock.MockClient lines result in go vet complaining about "mock.MockStore composite literal uses unkeyed fields" - which doesnt appear to be reality to me. Maybe go vet confuses slice literal with omitted types? Im not sure how to resolve those fails.

**Comment by carmstrong** [comment]
@iancoffey Looks like this needs a rebase.

**Comment by mboersma** [comment]
> the mock.MockClient lines result in go vet complaining...

Yep, and there isn't a workaround AFAICT: golang/go#9171 :disappointed:

**Comment by iancoffey** [comment]
fwiw - I tried the go vet workarounds we talked about (as far as I understood them) and had the same unfortunate result.

**Comment by carmstrong** [comment]
Can we have `go vet` ignore that line?

**Comment by iancoffey** [comment]
this I do not know - i havent found a nice way to do that yet without skipping all the structtags

**Comment by iancoffey** [comment]
This should go green now :four_leaf_clover:

File: docs/managing_deis/upgrading-deis.rst [hunk 6]
```diff
@@ -63,6 +63,56 @@ Finally, update ``deisctl`` to the new version and reinstall:
     enable PROXY protocol for ports 80 and 443 on the ELB, add a ``TCP 443:443`` listener, and
     change existing targets and health checks from HTTP to TCP.
 
+
+Graceful Upgrade
```
Comment (krancour) [inline docs/managing_deis/upgrading-deis.rst]: @iancoffey since Deis < 1.9.0 won't support `deisctl upgrade-prep` and 1.9.0 will if this is merged, the instructions documented here won't actually _work_ for anyone until upgrading from m >= 1.9.0 to n > m.

Would it make sense to _not_ include this documentation until the release after this feature is rolled out?  Otherwise we may have people attempting an upgrade procedure that is guaranteed to fail.

Comment (krancour) [inline docs/managing_deis/upgrading-deis.rst]: I wouldn't throw the docs away either-- just break them out into another PR that we can sit on for a while.

Comment (carmstrong) [inline docs/managing_deis/upgrading-deis.rst]: Good point. I think we could just add a note here stating that this can be used for 1.9.0 moving forward.

Comment (krancour) [inline docs/managing_deis/upgrading-deis.rst]: :+1:

Comment (krancour) [inline docs/managing_deis/upgrading-deis.rst]: Should it also be noted somewhere that this won't work across major versions either?

File: deisctl/config/api.go [hunk 7]
```diff
@@ -1,8 +0,0 @@
-package config
```
Comment (krancour) [inline deisctl/config/api.go]: @iancoffey I'm not sure about relocating this interface to the `etcdclient` package.  This interface looks like / was previously a generic "config client" interface of which many implementations are possible-- your new mock implementation of that being one of them.  But moving this interface to the `etcdclient` means users of the interface will be coding to an abstraction (the interface) whose package name is polluted with the name of a specific underlying implementation.

Was there a specific reason for moving it?

File: deisctl/cmd/upgrade.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	b.Destroy([]string{"store-daemon"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	b.Stop([]string{"store-monitor"}, &wg, Stdout, Stderr)
+	wg.Wait()
+	b.Destroy([]string{"store-monitor"}, &wg, Stdout, Stderr)
+	wg.Wait()
+
+	fmt.Fprintln(Stdout, "The platform has been stopped, but applications are still serving traffic as normal.")
+	fmt.Fprintln(Stdout, "Your cluster is now ready for upgrade. Install a new deisctl version and run deisctl take-over.")
```
Comment (krancour) [inline deisctl/cmd/upgrade.go]: @iancoffey Doesn't match the actual command that does the take-over.  Should be `upgrade-takeover`.

Also, for maximum clarity and consistency, you might consider using backticks to demarcate the boundaries of the command you're asking them to run.  That's done elsewhere in deisctl.  As a for instance, this is the message provided after a `deisctl install platform` reads as follows:

```
Please run `deisctl start platform` to boot up Deis.
```

File: docs/managing_deis/upgrading-deis.rst [hunk 9]
```diff
@@ -63,6 +63,56 @@ Finally, update ``deisctl`` to the new version and reinstall:
...
+----------------
+
+Alternatively, an experimental feature exists to provide the ability to perform a graceful upgrade. Unlike the
+in-place process above, this process keeps the platforms routers and publishers up during the upgrade process.
+This means that there should only be a maximum of around 1-2 seconds of downtime while the routers boot up. Many
```
Comment (krancour) [inline docs/managing_deis/upgrading-deis.rst]: > This means that there should only be a maximum of around 1-2 seconds of downtime while the routers boot up

Is this on account of how health checks from the load balancers are configured?  I can imagine a scenario where even with the rolling restarts, the timing of the health checks that determine a router is down and then back up could have all _n_ routers out of commission for a while.  This would happen if health checks were frequent and quick to consider a router down after a couple bad responses, but slow to consider it back up again by requiring some substantial number of healthy responses first.

File: deisctl/cmd/upgrade.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	if err := doUpgradeTakeOver(b, etcdClient); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func doUpgradeTakeOver(b backend.Backend, etcdClient etcdclient.Client) error {
```
Comment (krancour) [inline deisctl/cmd/upgrade.go]: @iancoffey should this function also update the value of `/deis/platform/version` in etcd?  Otherwise, won't the same versions of each component just come right back up (despite a newer version of deisctl having been used to do so)?

Referencing the inner-workings of the `get_image` function:

```
#!/bin/bash
# usage: get_image <component_path>
IMAGE=`etcdctl get $1/image 2>/dev/null`

# if no image was set in etcd, we use the default plus the release string
if [ $? -ne 0 ]; then
  RELEASE=`etcdctl get /deis/platform/version 2>/dev/null`

  # if no release was set in etcd, use the default provisioned with the server
  if [ $? -ne 0 ]; then
    source /etc/deis-release
    RELEASE=$DEIS_RELEASE
  fi

  IMAGE=$1:$RELEASE
fi

# remove leading slash
echo ${IMAGE#/}
```

Come to think of it, it needs to be noted that the upgrade won't upgrade any components for which a custom image has already been specified via etcd key `/deis/<component>/image`

**Comment by krancour** [comment]
@iancoffey despite my nitpicking... I think this is great!  Locally, this worked just as advertised and the only downtime I experienced was when the router hosted on `172.17.8.100` recycled and that was to be expected since there's no LB in play and `*.local3.deisapp.com` just points to that IP.

**Comment by mboersma** [comment]
This did pass despite the mixed checks indicated above: https://ci.deis.io/job/test-ec2/2417/

File: deisctl/cmd/upgrade.go (hunk 10 above)
Comment (carmstrong) [inline deisctl/cmd/upgrade.go]: I think we ask the user to explicitly set the new platform version in the Graceful Upgrade docs.

Comment (iancoffey) [inline deisctl/cmd/upgrade.go]: Yeah, setting the version is part of the procedure. IMO thats fine for the time being.

Comment (krancour) [inline deisctl/cmd/upgrade.go]: I see that now.  If the consensus is that's sufficient, I'm happy to go along with it, but for my own edification, why would't we want to set that for them as part of `deisctl upgrade-takeover`?

File: docs/managing_deis/upgrading-deis.rst (hunk 9 above)
Comment (iancoffey) [inline docs/managing_deis/upgrading-deis.rst]: This is based on the actual availability of the routers - there will almost certainly be at least one functional router up at all times. I originally had some text about configuring LBs to avoid that second or two of downtime, but it ended up being a bit much. Ill re-add something short and generic to explain the concern.

Comment (krancour) [inline docs/managing_deis/upgrading-deis.rst]: > Ill re-add something short and generic to explain the concern

Got it.  My only concern was that the following sounds a bit non-deterministic, so a brief explanation would help:

> This means that there should only be a maximum of around 1-2 seconds of downtime while the routers boot up. Many times, there will be no downtime at all.

File: deisctl/cmd/upgrade.go (hunk 10 above)
Comment (iancoffey) [inline deisctl/cmd/upgrade.go]: To me, its just a clearer workflow and less code in this already huge PR. We already have the logic to do this without more magic - I suggest we start there and add more logic down the road. I can create an issue if you want to keep track of that. Otherwise, Ill defer to maintainers consensus on if its worth it - Ill gladly add it if it is.

Comment (krancour) [inline deisctl/cmd/upgrade.go]: Good enough for me.

File: docs/managing_deis/upgrading-deis.rst (hunk 6 above)
Comment (iancoffey) [inline docs/managing_deis/upgrading-deis.rst]: Yes indeed.

File: deisctl/config/api.go (hunk 7 above)
Comment (iancoffey) [inline deisctl/config/api.go]: Yeah - I likely confused the config interface with an etcd specific interface since the terms are similar.

**Comment by krancour** [comment]
Replaced by #4175
